### ISSUE TITLE
Incompatible types when nesting anonymous structs

### ISSUE THREAD

**Issue by andyjansson** [UbiquityDotNET/Llvm.NET#192]
Hi, 

I switched to alpha.3 after I noticed that you had published it and promptly made some progress on the code generation. I can create anonymous structs just fine now, however, when trying to nest structs I'm faced with the following exception:

```
Incompatible types: destination pointer must be of the same type as the value stored.
```

I'm using the visitor pattern to traverse my AST, generating IR as I go. When I visit a node, I translate that node's type to its corresponding LLVM type. It seems like Llvm.NET is not satisfied with that, requiring the exact type. Is there any way around this?

This is my code: 

```csharp
public override Expression VisitTupleExpression(TupleExpression node)
{
     var type = GetType(node.Type);
     var tuplePtr = builder.Alloca(type);

     for (var i = 0; i < node.Elements.Count(); i++)
     {
          var element = node.Elements.ElementAt(i);
          Visit(element);
          var value = valueStack.Pop();
          var valuePtr = builder.GetStructElementPointer(type, tuplePtr, (uint)i);
          builder.Store(value, valuePtr);
     }

     valueStack.Push(tuplePtr);
     return node;
}
```

**Comment by smaillet** [comment]
I'm not really sure of the types of things in that sample. (When posting a sample snippet of code to a public place like GitHub it helps to limit the use of var [or auto in C++] and use explicit types as the context for them is usually lost in a snippet). It isn't clear if value which is popped from your stack is actually of the correct type to match the structure element. (Might want to put in an assert on that point in the code)

Can you reduce the problem to a simple small sample case (Ideally logging details of the types via ToString() calls sent to the console or debug output)?

It is worth noting, that InstructionBuilder.Store() requires that value.NativeType == valuePtr.ElementType AND for integeral types that includes matching the bit width of the integer. (The checks are done in managed code before calling the underlying LLVMBuildStore native function as LLVM doesn't do any param checks and will rather rudely just abort the process rather unceremoniously with little in the way of diagnostic output to explain why)

**Comment by andyjansson** [comment]
Hi, 

The snippet was meant to illustrate how I'm building the IR, not to provide a sample case. What I was trying to illustrate is that the tuple type doesn't have access to its members' types and vice versa. 
The point is that `value.ElementType` isn't the same as `valuePtr.NativeType`. They're *identical*, but they're *not the same object*. The issue is that I'm seemingly blocked by reference equality.

**Comment by andyjansson** [comment]
Another issue which seems related to this is in `InstructionBuilder.Return`. Suppose I have a function which returns an anonymous 0-tuple `fun foo() = ();`. Since the `IrFunction` is created beforehand (for two reasons 1. to it may be referenced by other functions, and 2. because we need to attach the entry block to something) and because the tuple is unnamed, we cannot know what the exact type of the return value is going to be.  Therefore we'll be hit with the following exception: 

```
Exception thrown: 'System.ArgumentException' in Ubiquity.NET.Llvm.dll: 'Value for return must match the function signature's return type'
```

Again, this is due to relying on [reference equality rather than structural equality](https://github.com/UbiquityDotNET/Llvm.NET/blob/develop/src/Ubiquity.NET.Llvm/Instructions/InstructionBuilder.cs#L341-L344).

**Comment by smaillet** [comment]
I think there's some confusion on how the underlying LLVM type system works and how the LLVM.NET libraries are designed to reflect it. The reference equality isn't a bug or design flaw, it is fundamentally required by LLVM itself, really. the LLVM.NET bindings go to great lengths to ensure that behavior operates consistently with how LLVM operates. note: the LL in LLVM is "Low Level". LLVM IR is not a high level language by any stretch and has to support the semantics of many different languages.

From the [LLVM docs](https://releases.llvm.org/10.0.0/docs/LangRef.html#structure-types):
>LLVM IR allows you to specify both “identified” and “literal” structure types. Literal types are uniqued structurally, but identified types are **never uniqued**. An opaque structural type can also be used to forward declare a type that is not yet available. [Emphasis mine]

Since identified structural types are NOT uniqued, two structs, even with the same type layout, are considered distinct types. Or in other words just because two types are layout compatible, doesn't mean they are semantically compatible. It is on the code generator to ensure such compatibility. Since the code generator knows the context and rules of the language it can consider structural equality and inject a bitcast to let the LLVM type system know they are considered equivalent. 

As to declaring the function without knowing the type, I'm not familiar with the nuances of your use case where that is required, generally the return type of a function is known by the front end as it is part of the signature/declaration of the function. Even C++ `decltype` is just compiler frontend sugar for the actual type that is known. The compiler frontend needs to resolve such things to the actual type before declaring the function.

**Comment by andyjansson** [comment]
My front-end does know the types of its program. I'm using a structural type system which has its typed inferenced using Hindley-Milner. Because it is structural and not nominal, I can't ask LLVM's type system to return a type of a previously declared type by name. What I meant by not knowing the type isn't that the program isn't type checked, but that the codegen can't know the body's LLVM type until it's made a pass over that AST node and exchanged its internal type for it - same goes for other subtrees. 

>Since the code generator knows the context and rules of the language it can consider structural equality and inject a bitcast to let the LLVM type system know they are considered equivalent.

Great, bitcasting seems to be what I want. Thanks.
